### PR TITLE
Updated Admin CLI commands to support tenants.

### PR DESCRIPTION
**Pull request by msterin**
Defines the syntax for the following `vmdkops_admin.py` commands: 
```
tenant {create, rm , ls} - manipulates tenants
tenant vm {add, rm, ls}  - manipulates VMs for a tenant
tenant access {add, set, rm, ls} - manipulates datastore access right for a tenant
```

Note that the actual commands implementation is still TBD - this change only defines
new command syntax and keywords. When invoked with correct arguments, the commands simply print `Not implemented`.

//CC @andrewjstone @lipingxue @marksoper 

Note: doc update is tracked in #622

TESTED: manually , by asking for `--help`. Log below
```
[root@localhost:~] 
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py  status 
Version: 0.7.3eb1146-0.0.1
Status: Running
Pid: 51662
Port: 1019
LogConfigFile: /etc/vmware/vmdkops/log_config.json
LogFile: /var/log/vmware/vmdk_ops.log
LogLevel: DEBUG
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant
usage: vmdkops_admin.py tenant [-h] {access,rm,create,ls,vm} ...
vmdkops_admin.py tenant: error: too few arguments
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant create --help 
usage: vmdkops_admin.py tenant create [-h] [--vm-list vm1, vm2, ...] name

positional arguments:
  name                  The name of the tenant

optional arguments:
  -h, --help            show this help message and exit
  --vm-list vm1, vm2, ...
                        A list of VM names to place in this Tenant
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant rm --help 
usage: vmdkops_admin.py tenant rm [-h] [--remove-volumes REMOVE_VOLUMES] name

positional arguments:
  name                  The name of the tenant

optional arguments:
  -h, --help            show this help message and exit
  --remove-volumes REMOVE_VOLUMES
                        BE CAREFUL - Removes this tenant volumes when removing
                        a tenant
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant ls --help 
usage: vmdkops_admin.py tenant ls [-h]

optional arguments:
  -h, --help  show this help message and exit
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant vm --help 
usage: vmdkops_admin.py tenant vm [-h] {rm,add,ls} ...

positional arguments:
  {rm,add,ls}
    rm         Remove VM(s) from a tenant
    add        Add a VM(s) to a tenant
    ls         list VMs in a tenant

optional arguments:
  -h, --help   show this help message and exit
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant vm add --help 
usage: vmdkops_admin.py tenant vm add [-h] --name NAME --vm-list VM_LIST

optional arguments:
  -h, --help         show this help message and exit
  --name NAME        Tenant to add the VM to
  --vm-list VM_LIST  A list of VM names to add to this Tenant
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant vm rm --help 
usage: vmdkops_admin.py tenant vm rm [-h] --name NAME --vm-list VM_LIST

optional arguments:
  -h, --help         show this help message and exit
  --name NAME        Tenant to remove the VM from
  --vm-list VM_LIST  A list of VM names to rm from this Tenant
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant vm ls --help 
usage: vmdkops_admin.py tenant vm ls [-h] [--name NAME]

optional arguments:
  -h, --help   show this help message and exit
  --name NAME  Tenant to list the VMs for
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant access --help 
usage: vmdkops_admin.py tenant access [-h] {rm,add,set,ls} ...

positional arguments:
  {rm,add,set,ls}
    rm             Remove all access to a datastore for a tenant
    add            Add a datastore access for a tenant
    set            Modify datastore access for a tenant
    ls             List all access info for a tenant

optional arguments:
  -h, --help       show this help message and exit
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant access add --help 
usage: vmdkops_admin.py tenant access add [-h] --name NAME
                                          [--volume-maxsize Num{MB,GB,TB} - e.g. 2TB]
                                          --datastore DATASTORE
                                          [--rights create,delete,mount]

optional arguments:
  -h, --help            show this help message and exit
  --name NAME           The name of the tenant
  --volume-maxsize Num{MB,GB,TB} - e.g. 2TB
                        Maximum size of the volume that can be created
  --datastore DATASTORE
                        Datastore which access is controlled
  --rights create,delete,mount
                        Datastore access Permissions granted: Choices =
                        ['create', 'delete', 'mount', 'all']
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant access rm  --help 
usage: vmdkops_admin.py tenant access rm [-h] --name NAME --datastore
                                         DATASTORE

optional arguments:
  -h, --help            show this help message and exit
  --name NAME           The name of the tenant
  --datastore DATASTORE
                        Datstore which access is controlled
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant access set --help 
usage: vmdkops_admin.py tenant access set [-h]
                                          [--rm-rights create,delete,mount]
                                          --name NAME
                                          [--volume-maxsize Num{MB,GB,TB} - e.g. 2TB]
                                          [--add-rights create,delete,mount]
                                          --datastore DATASTORE

optional arguments:
  -h, --help            show this help message and exit
  --rm-rights create,delete,mount
                        Datastore access Permissions granted: Choices =
                        ['create', 'delete', 'mount', 'all']
  --name NAME           Tenant name
  --volume-maxsize Num{MB,GB,TB} - e.g. 2TB
                        Maximum size of the volume that can be created
  --add-rights create,delete,mount
                        Datastore access Permissions granted: Choices =
                        ['create', 'delete', 'mount', 'all']
  --datastore DATASTORE
                        Datastore name
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant access ls --help 
usage: vmdkops_admin.py tenant access ls [-h] --name NAME

optional arguments:
  -h, --help   show this help message and exit
  --name NAME  The name of the tenant
```

### CODE REVIEW

**Comment by pdhamdhere** [comment]
@msterin I think you need to fix tests to remove "role" subcommand. For DS access, I only see maxVolSize. Does underlying API support maxNumVols or quota-Per-Datastore?

**Comment by msterin** [comment]
@pdhamdhere  - yes, that's what happens in a rush :-). Thanks for the review, I will fix tests/ add numvol/quota and refresh the PR

File: docs/misc/vmdkops-admin-cli-spec.md [hunk 1]
```diff
@@ -11,7 +11,9 @@ All output from the admin cli defaults to human readable formats. It will be mad
 The majority of testing will be automated. We can ensure that parsing calls the right callbacks with
 the right information by generating representative input and mocking the callbacks to assert that
 the right information is parsed and delivered correctly. Additionally, and specifically for testing
-roles, we can create roles and then test that they act as expected by calling `role get`. Unit
+access control, we can create access controk definition (tenants and privileges)
```
Comment (govint) [inline docs/misc/vmdkops-admin-cli-spec.md]: nit - "control"

File: docs/misc/vmdkops-admin-cli-spec.md [hunk 2]
```diff
@@ -11,7 +11,9 @@ All output from the admin cli defaults to human readable formats. It will be mad
 The majority of testing will be automated. We can ensure that parsing calls the right callbacks with
 the right information by generating representative input and mocking the callbacks to assert that
 the right information is parsed and delivered correctly. Additionally, and specifically for testing
-roles, we can create roles and then test that they act as expected by calling `role get`. Unit
+access control, we can create access controk definition (tenants and privileges) 
+and then test that they act as expected by invoking vmdk_ops commmands on behalf of a fake VM.
```
Comment (govint) [inline docs/misc/vmdkops-admin-cli-spec.md]: Can suggest using the dummy VM created with vim-cmd and say how that can be created.

File: esx_service/cli/vmdkops_admin.py [hunk 3]
```diff
@@ -94,20 +94,20 @@ def commands():
...
     * cmds - A dictionary of subcommands where the key is the next word in the command line string.
-             For example, in `vmdkops_admin.py role create`, `role` is the command, and `create` is
+             For example, in `vmdkops_admin.py tenant create`, `tenant` is the command, and `create` is
              the subcommand. Subcommands can have further subcommands, but currently there is only
              one level of subcommands in this specification. Each subcommand can contain the same
              attributes as top level commands: (func, help, args, cmds). These attributes have
              identical usage to the top-level keys, except they only apply when the subcommand is
-             part of the command. For example the `--matches-vm` argument only applies to `role
-             create` or `role set` commands. It will be invalid in any other context.
+             part of the command. For example the `--matches-vm` argument only applies to `tenant
```
Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: The --matches-vm option seems to be removed. Remove it from here as well.

File: esx_service/cli/vmdkops_admin.py [hunk 4]
```diff
@@ -121,6 +121,9 @@ def commands():
                                 'attached-to', 'policy', 'capacity', 'used',
                                 'fstype', 'access', 'attach-as'],
                     'metavar': 'Col1,Col2,...'
+                },
+                '--tenant' : {
+                    'help': 'Displays only VMs for a given tenant'
```
Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: Are there any entities to display besides VMs, if not then can reword as "Displays VMs for a given tenant".

Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: Should --tenant be a part of the ls command? It seems to be listing volumes by default.

Instead we could have like:

vmdkops_admin with commands like
a. volume - ls, create, rm, set
b. tenant - ls, create, rm, set
c. <anymore>

which matches with esxcli type command options where we have broad categories within which the same CRUD options are available (list, create, delete, ....)

File: esx_service/cli/vmdkops_admin.py [hunk 5]
```diff
@@ -172,83 +175,186 @@ def commands():
...
+                    'help': 'Delete a tenant',
                     'args': {
                         'name': {
-                            'help': 'The name of the role'
-                        }
+                            'help': 'The name of the tenant',
+                      },
+                      '--remove-volumes': {
+                        'help': 'BE CAREFUL -  Removes this tenant volumes when removing a tenant'
+                      }
```
Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: Or something like "WARNING: - Removes this tenant's volumes from all datastores when removing a tenant

File: esx_service/cli/vmdkops_admin.py [hunk 6]
```diff
@@ -172,83 +175,186 @@ def commands():
...
-                            'required': True,
-                            'metavar': 'Num{MB,GB,TB} - e.g. 2TB'
                         }
                     }
                 },
                 'rm': {
-                    'func': role_rm,
-                    'help': 'Delete a role',
+                    'func': tenant_rm,
+                    'help': 'Delete a tenant',
```
Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: "Remove a tenant" as its tenant_rm vs. tenant_delete?

File: esx_service/cli/vmdkops_admin.py [hunk 7]
```diff
@@ -172,83 +175,186 @@ def commands():
...
+                                    'help': "Datastore name",
+                                    'required': True
+                                },
+                                '--add-rights': {
+                                    'help': 'Datastore access Permissions granted',
+                                    'choices': ['create', 'delete', 'mount', 'all'],
+                                    'metavar': 'create,delete,mount'
+                                }, 
+                                '--rm-rights': {
+                                    'help': 'Datastore access Permissions granted',
```
Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: "Datastore access permissions removed"

Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: Also, rm-rights must allow a way to specify a list of datastores or "all" datastores in one go.

File: esx_service/cli/vmdkops_admin.py [hunk 8]
```diff
@@ -172,83 +175,186 @@ def commands():
...
+                                }
+                            }
+                        },
+                        'rm': {
+                            'func': tenant_access_rm,
+                            'help': "Remove all access to a datastore for a tenant",
+                            'args': {
+                                '--name': {
+                                    'help': 'The name of the tenant',
+                                    'required': True
```
Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: This should be a list, a use case would be user wants to retire a datastore. User should be able to get all tenants to stop using the datastore.

File: esx_service/cli/vmdkops_admin.py [hunk 9]
```diff
@@ -172,83 +175,186 @@ def commands():
...
+                                }
+                            }
+                        },
+                        'ls': {
+                            'func': tenant_access_ls,
+                            'help': 'List all access info for a tenant',
+                            'args': {
+                                '--name': {
+                                    'help': 'The name of the tenant',
+                                    'required': True
```
Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: Again, list for named tenants or make required as "False" and when not provided list access for all tenants.

File: esx_service/cli/vmdkops_admin.py [hunk 10]
```diff
@@ -324,7 +430,7 @@ def make_list_of_values(allowed):
     Take a list of allowed values for an option and return a function that can be
     used to typecheck a string of given values and ensure they match the allowed
     values.  This is required to support options that take comma seperated lists
-    such as --rights in 'role set --rights=create,delete,mount'
+    such as --rights in 'tenant set --rights=create,delete,mount'
```
Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: Reword this to mention --add-rights or --rm-rights for tenent access set command.

File: esx_service/cli/vmdkops_admin.py (hunk 4 above)
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: this command is supposed to do LS on volumes , filtered by a tenant.

File: esx_service/cli/vmdkops_admin.py (hunk 6 above)
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: yes, all commands are "rm"

File: esx_service/cli/vmdkops_admin.py [hunk 11]
```diff
@@ -172,83 +175,186 @@ def commands():
...
+                        'set': {
+                            'func': tenant_access_set,
+                            'help': 'Modify datastore access for a tenant',
+                            'args': {
+                                '--name': {
+                                    'help': 'Tenant name',
+                                    'required': True
+                                },
+                                '--datastore': {
+                                    'help': "Datastore name",
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: I agree but I am not sure how to configure it. @andrewjstone - any hints here ?

File: esx_service/cli/vmdkops_admin.py (hunk 10 above)
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: that is just an example in comments, it does not need to enumerate all usages

**Comment by msterin** [comment]
@govint - thanks for the review ! I replied to some of the comment and fixed the rest. Everything not replied to is fixed.

**Comment by pdhamdhere** [comment]
CC/ @lipingxue Can you please review?
